raulin0/backup-juggler
# feat: adds the backup funcionality

# Description
This pull request introduces several enhancements to the Backup Juggler application. It adds the ability to generate documentation using MkDocs, updates the project's dependencies and configurations, adds unit tests for the backup functionality, includes fixtures for creating temporary paths during testing, implements a ProgressViewer class for displaying progress during file copying, and introduces a FileCopier class and backup functions for file copying operations. Additionally, a Paths model class is added to represent source and destination paths.

# Changes Made
- Added the mkdocs.yml configuration file to enable documentation generation using MkDocs.
- Updated the pyproject.toml file to include new dependencies, dev dependencies, and doc dependencies, as well as task automation configurations.
- Added unit tests for the backup functionality, covering different scenarios such as copying files and directories, handling non-existent source files, and comparing file sizes.
- Added test fixtures to facilitate the creation of temporary paths for testing purposes.
- Implemented the ProgressViewer class to display the progress of the file copying operation using a progress bar.
- Introduced the FileCopier class to handle file copying operations from source to destination.
- Implemented backup and parallel_backups functions to initiate backup operations.
- Added the Paths model class to represent source and destination paths.

# Motivation and Context
These changes aim to enhance the Backup Juggler application by adding documentation generation, improving testing capabilities, providing progress feedback during file copying, and introducing a more organized approach to file copying operations. The additions and updates contribute to the overall functionality, maintainability, and user experience of the application.

# How to Test
1. Ensure that the project dependencies and dev dependencies, including tqdm, pytest, and mkdocs-material, are installed.
2. Run the test suite to verify the existing and new functionality.
3. Perform backup operations with different scenarios, including copying files, directories, and nested subdirectories, and verify the correctness of the copied files and directories.
4. Run the documentation generation using MkDocs and verify that the documentation is generated successfully with the expected content.
5. Manually test the progress viewer by performing file copying operations and ensuring that the progress bar accurately reflects the progress.
6. Verify that the original functionality of the Backup Juggler application remains intact.

# Related Issue(s)
Include any relevant issue numbers or links here.

# Checklist

Please review and merge this pull request at your earliest convenience. Feedback and suggestions for improvement are welcome. Thank you!
